File: packages/shared/src/billing/records/worklog-list.ts
import type { WorklogRow } from '../types.js';

export interface WorklogDay {
  date: string;
  totalMinutes: number;
  entries: WorklogRow[];
}

export function groupWorklogsByDay(
  rows: WorklogRow[],
  opts: { month: string; projectId?: number },
): WorklogDay[] {
  const { month, projectId } = opts;
  const byDate = new Map<string, WorklogDay>();
  for (const r of rows) {
    if (r.workDate.slice(0, 7) !== month) continue;
    if (projectId !== undefined && r.projectId !== projectId) continue;
    const day = byDate.get(r.workDate) ?? { date: r.workDate, totalMinutes: 0, entries: [] };
    day.totalMinutes += r.minutes;
    day.entries.push(r);
    byDate.set(r.workDate, day);
  }
  return [...byDate.values()].sort((a, b) => (a.date < b.date ? 1 : a.date > b.date ? -1 : 0));
}

File: swift/WatchtowerCore/Sources/WatchtowerCore/Billing/records/WorklogList.swift
import Foundation

/// A single day's worklog entries — mirrors `packages/shared/src/billing/records/worklog-list.ts`.
public struct WorklogDay: Equatable, Sendable {
    public let date: String
    public let totalMinutes: Double
    public let entries: [WorklogRow]

    public init(date: String, totalMinutes: Double, entries: [WorklogRow]) {
        self.date = date
        self.totalMinutes = totalMinutes
        self.entries = entries
    }
}

/// Groups `rows` by `workDate` within `month` (YYYY-MM prefix), optionally
/// scoped to `projectId`. Sums raw `minutes` (not `effectiveMinutes`).
/// Entries keep encounter order within a day. Result is sorted descending
/// by date (newest day first).
public func groupWorklogsByDay(
    _ rows: [WorklogRow], month: String, projectId: Int?
) -> [WorklogDay] {
    var byDate: [String: WorklogDay] = [:]
    var order: [String] = []
    for r in rows {
        guard r.workDate.prefix(7) == month else { continue }
        if let projectId, r.projectId != projectId { continue }
        if byDate[r.workDate] == nil {
            byDate[r.workDate] = WorklogDay(date: r.workDate, totalMinutes: 0, entries: [])
            order.append(r.workDate)
        }
        let day = byDate[r.workDate]!
        byDate[r.workDate] = WorklogDay(
            date: day.date,
            totalMinutes: day.totalMinutes + r.minutes,
            entries: day.entries + [r]
        )
    }
    return order.map { byDate[$0]! }.sorted { $0.date > $1.date }
}
